crates/runtime/src/http/auth.rs

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use axum::{
+    body::Body,
+    http::StatusCode,
+    http::{Request, Response},
+};
+use futures::future::BoxFuture;
+use runtime_auth::{AuthVerdict, HttpAuth};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tower::{Layer, Service};
+
+#[derive(Clone)]
+pub struct AuthLayer {
+    auth_verifier: Arc<dyn HttpAuth + Send + Sync>,
+}
+
+impl AuthLayer {
+    pub fn new(auth_verifier: Arc<dyn HttpAuth + Send + Sync>) -> Self {
+        Self { auth_verifier }
+    }
+}
+
+impl<S> Layer<S> for AuthLayer {
+    type Service = AuthMiddleware<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AuthMiddleware {
+            inner,
+            auth_verifier: Arc::clone(&self.auth_verifier),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct AuthMiddleware<S> {
+    inner: S,
+    auth_verifier: Arc<dyn HttpAuth + Send + Sync>,
+}
+
+impl<S> Service<Request<Body>> for AuthMiddleware<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Response<Body>, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let auth_verifier = Arc::clone(&self.auth_verifier);
+        let mut inner = self.inner.clone();
+
+        Box::pin(async move {
+            let (parts, body) = req.into_parts();
+
+            match auth_verifier.http(&parts) {
+                Ok(AuthVerdict::Allow) => inner.call(Request::from_parts(parts, body)).await,
+                Ok(AuthVerdict::Deny) => Ok(unauthorized_response()),
+                Err(e) => {
+                    tracing::error!("{e}");
+                    Ok(internal_server_error_response())
+                }
+            }
+        })
+    }
+}
+
+fn unauthorized_response() -> Response<Body> {
+    match Response::builder()
+        .status(StatusCode::UNAUTHORIZED)
+        .body(Body::from("Unauthorized"))
+    {
+        Ok(response) => response,
+        Err(e) => panic!("Failed to build unauthorized response: {e}"),
```

Comment:
This should never happen and we can't return an error here because axum middlewares should be infallible. The correct way to handle an error is to return an error response, but since this IS the error response, there isn't much we can do if it can't be constructed. I will test this in future PRs to ensure it doesn't happen.